File: 064. Minimum Path Sum/minPathSum.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function(grid) {
    const [row,column] = [grid.length,grid[0].length]
    for (let i=0;i<row;i++) {
        for (let j=0;j<column;j++) {
            if (i==0 & j==0) {
                grid[i][j] += 0
            }
            else if (i==0) {
                grid[i][j] += grid[i][j-1]
            }
            else if (j==0) {
                grid[i][j] += grid[i-1][j]
            }
            else {
                grid[i][j] += Math.min(grid[i-1][j], grid[i][j-1])
            }
        }
    }
    return grid[row-1][column-1]
};

// Runtime 105 ms / Memory 42.8 MB
// debug: https://jsfiddle.net/

File: 064. Minimum Path Sum/minPathSum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        // dp
        int row = grid.size();
        int col = grid[0].size();

        for (int r=0; r<row; r++) {
            for (int c=0; c<col; c++) {
                // origin
                if (r==0 && c==0) { grid[r][c] = grid[r][c]; }
                // boundary initial
                else if (r==0) { grid[r][c] += grid[r][c-1]; }
                else if (c==0) { grid[r][c] += grid[r-1][c]; }
                // others
                else {  grid[r][c] += min(grid[r-1][c], grid[r][c-1]); }
            }
        }

        return grid[row-1][col-1];
    }
};